File: src/main/java/org/example/expert/domain/todo/repository/TodoQueryRepositoryImpl.java
package org.example.expert.domain.todo.repository;

import java.time.LocalDate;
import java.util.List;
import java.util.Optional;

import org.example.expert.domain.comment.entity.QComment;
import org.example.expert.domain.manager.entity.QManager;
import org.example.expert.domain.todo.dto.response.TodoSummaryResponse;
import org.example.expert.domain.todo.entity.QTodo;
import org.example.expert.domain.todo.entity.Todo;
import org.example.expert.domain.user.entity.QUser;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.PageImpl;
import org.springframework.data.domain.Pageable;
import org.springframework.stereotype.Repository;
import org.springframework.util.StringUtils;

import com.querydsl.core.types.Projections;
import com.querydsl.core.types.dsl.BooleanExpression;
import com.querydsl.jpa.impl.JPAQueryFactory;

import lombok.RequiredArgsConstructor;

@Repository
@RequiredArgsConstructor
public class TodoQueryRepositoryImpl implements TodoQueryRepository {

	private final JPAQueryFactory queryFactory;

	@Override
	public Optional<Todo> findByIdWithUser(Long todoId) {
		QTodo todo = QTodo.todo;
		QUser user = QUser.user;

		return Optional.ofNullable(
			queryFactory
				.selectFrom(todo)
				.leftJoin(todo.user, user).fetchJoin()
				.where(todo.id.eq(todoId))
				.fetchOne()
		);
	}

	@Override
	public Page<TodoSummaryResponse> searchTodosWithFilters(String keyword, String nickname, LocalDate startDate,
		LocalDate endDate, Pageable pageable) {
		QTodo todo = QTodo.todo;
		QUser user = QUser.user;
		QManager manager = QManager.manager;
		QComment comment = QComment.comment;

		List<TodoSummaryResponse> content = queryFactory
			.select(Projections.constructor(
				TodoSummaryResponse.class,
				todo.title,
				manager.countDistinct(),
				comment.count()
			))
			.from(todo)
			.leftJoin(todo.managers, manager)
			.leftJoin(manager.user, user)
			.leftJoin(todo.comments, comment)
			.where(
				titleContains(keyword),
				nicknameContains(nickname),
				createdBetween(startDate, endDate)
			)
			.groupBy(todo.id)
			.orderBy(todo.createdAt.desc())
			.offset(pageable.getOffset())
			.limit(pageable.getPageSize())
			.fetch();

		Long count = queryFactory
			.select(todo.countDistinct())
			.from(todo)
			.leftJoin(todo.managers, manager)
			.leftJoin(manager.user, user)
			.where(
				titleContains(keyword),
				nicknameContains(nickname),
				createdBetween(startDate, endDate)
			)
			.fetchOne();

		return new PageImpl<>(content, pageable, count);
	}

	private BooleanExpression titleContains(String keyword) {
		return StringUtils.hasText(keyword) ? QTodo.todo.title.contains(keyword) : null;
	}

	private BooleanExpression nicknameContains(String nickname) {
		return StringUtils.hasText(nickname) ? QManager.manager.user.nickname.contains(nickname) : null;
	}

	private BooleanExpression createdBetween(LocalDate start, LocalDate end) {
		if (start != null && end != null) {
			return QTodo.todo.createdAt.between(start.atStartOfDay(), end.plusDays(1).atStartOfDay());
		}
		return null;
	}
}

File: src/main/kotlin/com/example/domain/todo/repository/TodoQueryRepositoryImpl.kt
package com.example.domain.todo.repository

import com.querydsl.core.types.Projections
import com.querydsl.core.types.dsl.BooleanExpression
import com.querydsl.jpa.impl.JPAQueryFactory
import com.example.domain.comment.entity.QComment
import com.example.domain.manager.entity.QManager
import com.example.domain.todo.dto.response.TodoSummaryResponse
import com.example.domain.todo.entity.QTodo
import com.example.domain.todo.entity.Todo
import com.example.domain.user.entity.QUser
import org.springframework.data.domain.Page
import org.springframework.data.domain.PageImpl
import org.springframework.data.domain.Pageable
import org.springframework.stereotype.Repository
import java.time.LocalDate
import java.util.*

@Repository
class TodoQueryRepositoryImpl(
    private val queryFactory: JPAQueryFactory
) : TodoQueryRepository {

    override fun findByIdWithUser(todoId: Long): Optional<Todo> {
        val todo = QTodo.todo
        val user = QUser.user

        return Optional.ofNullable(queryFactory
            .selectFrom(todo)
            .leftJoin(todo.user, user).fetchJoin()
            .where(todo.id.eq(todoId))
            .fetchOne()
        )
    }

    override fun searchTodosWithFilters(
        keyword: String?,
        nickname: String?,
        startDate: LocalDate?,
        endDate: LocalDate?,
        pageable: Pageable
    ): Page<TodoSummaryResponse> {
        val todo = QTodo.todo
        val user = QUser.user
        val manager = QManager.manager
        val comment = QComment.comment

        val content = queryFactory
            .select(
                Projections.constructor(
                    TodoSummaryResponse::class.java,
                    todo.title,
                    manager.countDistinct(),
                    comment.count()
                )
            )
            .from(todo)
            .leftJoin(todo.managers, manager)
            .leftJoin(manager.user, user)
            .leftJoin(todo.comments, comment)
            .where(
                titleContains(keyword),
                nicknameContains(nickname),
                createdBetween(startDate, endDate)
            )
            .groupBy(todo.id)
            .orderBy(todo.createdAt.desc())
            .offset(pageable.offset)
            .limit(pageable.pageSize.toLong())
            .fetch()

        val count = queryFactory
            .select(todo.countDistinct())
            .from(todo)
            .leftJoin(todo.managers, manager)
            .leftJoin(manager.user, user)
            .where(
                titleContains(keyword),
                nicknameContains(nickname),
                createdBetween(startDate, endDate)
            )
            .fetchOne() ?: 0L

        return PageImpl(content, pageable, count)
    }

    private fun titleContains(keyword: String?): BooleanExpression? {
        return if (!keyword.isNullOrBlank()) QTodo.todo.title.contains(keyword) else null
    }

    private fun nicknameContains(nickname: String?): BooleanExpression? {
        return if (!nickname.isNullOrBlank()) QManager.manager.user.nickname.contains(nickname) else null
    }

    private fun createdBetween(start: LocalDate?, end: LocalDate?): BooleanExpression? {
        return if (start != null && end != null) {
            QTodo.todo.createdAt.between(start.atStartOfDay(), end.plusDays(1).atStartOfDay())
        }  else null
    }
}
